File: src/main/java/wanted/Q01/Main.java
package wanted.Q01;

import java.util.Scanner;

public class Main {
    private static int min = Integer.MAX_VALUE;

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] c = new int[n + 1];
        for (int i = 1; i < n + 1; i++) {
            c[i] = in.nextInt();
        }
        int[][][] p = new int[n + 1][n - 1][2];
        for (int i = 1; i <= n; i++) {
            int m = in.nextInt();
            int[][] temp = new int[n - 1][2];
            for (int j = 0; j < m; j++) {
                temp[j][0] = in.nextInt();
                temp[j][1] = in.nextInt();
            }
            p[i] = temp;
        }
        Main main = new Main();
        main.solution(c, p);
        System.out.println(min);
    }

    private void solution(int[] c, int[][][] p) {
        int[] check = new int[c.length];
        check[0] = 1;
        DFS(0, 0, check, c, p);
    }

    private void DFS(int L, int sum, int[] check, int[] c, int[][][] p) {
        if (L == c.length - 1) {
            boolean allVisit = true;
            for (int i = 1; i < check.length; i++) {
                if (check[i] == 0) {
                    allVisit = false;
                }
            }
            if (allVisit) {
                min = Math.min(min, sum);
            }
        } else {
            for (int i = 1; i < c.length; i++) {
                if (check[i] == 0) {
                    int[][] pArr = p[i];
                    int[] copy = new int[c.length];
                    for (int i1 = 0; i1 < pArr.length; i1++) {
                        int[] temp = pArr[i1];
                        int idx = temp[0];
                        int discount = temp[1];
                        if (check[idx] == 1 || idx == i || discount == 0)
                            continue;
                        copy[idx] = c[idx];
                        c[idx] = c[idx] - discount <= 0 ? 1 : c[idx] - discount;
                    }
                    check[i] = 1;
                    DFS(L + 1, sum + c[i], check, c, p);
                    check[i] = 0;
                    for (int j = 0; j < copy.length; j++) {
                        if (copy[j] != 0) {
                            c[j] = copy[j];
                        }
                    }
                }
            }
        }
    }
}

File: src/main/java/wanted/Q01/main.py
def DFS(L, sum, check, cost, arr, answer):
    if L == (len(cost) - 1):
        all_visit = True
        for a in range(len(check)):
            if check[a] == 0:
                all_visit = False
        if all_visit:
            answer.append(sum)
    else:
        for a in range(1, len(cost)):
            if check[a] == 0:
                discount_map = arr[a]
                copy = [0 for k in range(len(cost))]
                for discount_info in discount_map:
                    idx = discount_info[0]
                    discount = discount_info[1]
                    if (check[idx] == 1) | (idx == a) | (discount == 0):
                        continue
                    copy[idx] = cost[idx]
                    cost[idx] = (cost[idx] - discount)
                    if cost[idx] <= 0:
                        cost[idx] = 1
                check[a] = 1
                DFS(L + 1, sum + cost[a], check, cost, arr, answer)
                check[a] = 0
                for b in range(len(copy)):
                    if copy[b] != 0:
                        cost[b] = copy[b]


def solution(cost, arr):
    check = [0 for i in range(len(cost))]
    check[0] = 1
    answer = []
    DFS(0, 0, check, cost, arr, answer)
    return min(answer)


n = int(input())
cost = list(map(int, input().split()))
cost.insert(0, 0)
arr = [[]]

for i in range(n):
    m = int(input())
    a = []
    for j in range(0, m):
        temp = list(map(int, input().split()))
        a.append(temp)
    arr.append(a)
answer = solution(cost, arr)
print(answer)
